Convert Dynamic Macro to a Core Feature (#5948)
qmk/qmk_firmware: docs/feature_dynamic_macros.md

You can store one or two macros and they may have a combined total of 128 keypresses. You can increase this size at the cost of RAM.

> To enable them, first include `DYNAMIC_MACRO_ENABLE = yes` in your Makefile. Then, add the following keys to your keymap:

Reviewer: ```suggestion
To enable them, first include `DYNAMIC_MACRO_ENABLE = yes` in your `rules.mk`. Then, add the following keys to your keymap:
```

?> For the details about the internals of the dynamic macros, please read the comments in the `process_dynamic_macro.h` and `process_dynamic_macro.c` files.

## Customization 

There are a number of options added that should allow some additional degree of customization

|Define                      |Default         |Description                                                                                                      |
|----------------------------|----------------|-----------------------------------------------------------------------------------------------------------------|
|`DYNAMIC_MACRO_SIZE         |128             |Sets the amount of memory that Dynamic Macros can use. This is a limited resource, dependent on the controller.  |
|`DYNAMIC_MACRO_USER_CALL`   |*Not defined*   |Defining this falls back to using the user `keymap.c` file to trigger the macro behavior.                        |

> If the LEDs start blinking during the recording with each keypress, it means there is no more space for the macro in the macro buffer. To fit the macro in, either make the other macro shorter (they share the same buffer) or increase the buffer size by adding the `DYNAMIC_MACRO_SIZE` define in your `config.h (default value: 128; please read the comments for it in the header).

Reviewer 2: Missing closing backtick after `` `config.h ``.

Author: Nice catch, thanks!

?> For the details about the internals of the dynamic macros, please read the comments in the `process_dynamic_macro.h` and `process_dynamic_macro.c` files.

## Customization 

There are a number of options added that should allow some additional degree of customization

|Define                      |Default         |Description                                                                                                      |
|----------------------------|----------------|-----------------------------------------------------------------------------------------------------------------|
|`DYNAMIC_MACRO_SIZE`        |128             |Sets the amount of memory that Dynamic Macros can use. This is a limited resource, dependent on the controller.  |
|`DYNAMIC_MACRO_USER_CALL`   |*Not defined*   |Defining this falls back to using the user `keymap.c` file to trigger the macro behavior.                        |
|`DYNAMIC_MACRO_NO_NESTING`  |*Not Defined*   |Defining this disables the ability to call a macro from another macro (nested macros).                           | 

> If the LEDs start blinking during the recording with each keypress, it means there is no more space for the macro in the macro buffer. To fit the macro in, either make the other macro shorter (they share the same buffer) or increase the buffer size by adding the `DYNAMIC_MACRO_SIZE` define in your `config.h (default value: 128; please read the comments for it in the header).

Reviewer 3: ```suggestion
If the LEDs start blinking during the recording with each keypress, it means there is no more space for the macro in the macro buffer. To fit the macro in, either make the other macro shorter (they share the same buffer) or increase the buffer size by adding the `DYNAMIC_MACRO_SIZE` define in your `config.h` (default value: 128; please read the comments for it in the header).
```

To enable them, first include `DYNAMIC_MACRO_ENABLE = yes` in your `rules.mk`. Then, add the following keys to your keymap:

|Key               |Alias     |Description                                        |
|------------------|----------|---------------------------------------------------|

> |`DYN_REC_START1`  |`DM_REC1` |Start recording Macro 1                            |

Reviewer: Could you add this to `keycodes.md` as well?

Reviewer 3: Oops. :rofl:

Reviewer 3: I think I'll do this in another PR.

Reviewer: Pushed to the wrong branch? 😄

Author: @noroadsleft push it to my branch again :D

Reviewer 3: @fauxpark Forgot that they needed to go into the keycodes Doc as well.

@drashna K, will do.

	}
```

### User Hooks

There are a number of hooks that you can use to add custom functionality and feedback options to Dynamic Macro feature.  This allows for some additional degree of customization. 

> Note, that direction indicates which macro it is, with `1` being Macro 1, `-1` being Macro 2, and 0 being no macro. 

Reviewer 2: Surround `0` in backticks to match `1` and `-1`.

Author: Thanks!  

Just a heads up, if you hit control+G, it lets you make a suggestion that can be committed by myself or others directly on github. 

Eg:
```suggestion
Note, that direction indicates which macro it is, with `1` being Macro 1, `-1` being Macro 2, and `0` being no macro. 
```

This isn't necessary to do (as I've already addressed the issues), but may be nice to know, in the future.
